fluent/client/ws_client.go

```diff
@@ -117,10 +145,19 @@ func (c *WSClient) Connect() error {
 		return err
 	}
 
-	c.Session = &WSSession{
-		ServerAddress: c.ServerAddress,
-		Connection:    connection,
-	}
+	c.Session = c.ConnectionFactory.NewSession(connection)
+	waitChan := make(chan struct{}, 1)
```

Comment:
Above function will not return the error if `c.Session.Connection.Listen()` returns an error. can be re written as below

```
func (c *WSClient) Connect() error {
	waitChan := make(chan struct{}, 1)
	defer func() {
		close(waitChan)
	}()
	var listenErr error

	if c.Session != nil {
		return errors.New("A session is already active")
	}

	if c.ConnectionFactory == nil {
		c.ConnectionFactory = &DefaultWSConnectionFactory{
			ServerAddress: c.ServerAddress,
			AuthInfo:      c.AuthInfo,
		}
	}

	conn, err := c.ConnectionFactory.New()
	if err != nil {
		return err
	}

	connection, err := ws.NewConnection(conn, c.ConnectionOptions)
	if err != nil {
		return err
	}

	c.Session = c.ConnectionFactory.NewSession(connection)

	go func() {
		if err := c.Session.Connection.Listen(); err != nil {
			listenErr = err
			waitChan <- struct{}{}
		} else {
			waitChan <- struct{}{}
		}
	}()

	// wait for go routine to start
	<-waitChan

	return listenErr
}
```

Reply:
> The channel `waitChan` will not be closed, so (formally) this will become a memory leak. Why is it buffered? I assume to not block the concurrent function during startup?!
> 
> This can also be achieved without memory leak:
> 
> ```
> waitChan := make(chan struct{})
> go func() {
>    close(waitChan)
>    .
>    .
> }
> 
> <- waitChan
> ```

You don't need to explicitly close channels unless the it's important to the control flow. E.g., the receiver may need to know the channel has been closed so it can exit a loop. In either case, the garbage collector will close the channel when it's no longer needed.

Reply:
I'm also adding comments to explain

Reply:
So, as you mentioned `error from Listen does matter to Connect ` then it's ok to start listen asynchronously and we can remove `waitChan `